transformer_engine/jax/cpp_extensions.py

```diff
@@ -1973,3 +1975,448 @@ def scaled_upper_triang_masked_softmax_bwd(grad_outputs: jnp.ndarray, softmax_ou
     return _scaled_upper_triang_masked_softmax_bwd_p.bind(grad_outputs,
                                                           softmax_outputs,
                                                           scale_factor=scale_factor)
+
+
+class SelfFusedAttnMax512FwdPrimitive(BasePrimitive):
+    """
+    Self Fused Attention Forward Primitive
+    """
+    name = "te_self_fused_attn_max_512_forward"
+    multiple_results = True
+
+    @staticmethod
+    def abstract(
+            qkv,
+            bias,
+            cu_seqlen,    # pylint: disable=unused-argument
+            rng_state,    # pylint: disable=unused-argument
```

Comment:
Why are these arguments included if unused? Same comment also in `sharding.py`?

Reply:
- For cpp_extension.py
This is [JAX custom call](https://jax.readthedocs.io/en/latest/Custom_Operation_for_GPUs.html#high-level-steps) required. People have to implement the abstract evaluation(`abstract`) and lowering rule(`lowering`) to MLIR. Both of them will be invoked by the JAX, so we have to follow it's format to define our arguments even we don't use them.

- For sharding.py
We pass all sharding information through [`get_fused_attn_sharding_meta`](https://github.com/NVIDIA/TransformerEngine/pull/175/files/74ee0e0483d87d0bbf187b68f7e20b0e4632a6fc#diff-1cc86ec7a8a0180afe58c84fbae92e7cfe8deb1fa9b2935a970b987c8b409bdaR1003), and [`ShardingMetaGenerator`](https://github.com/NVIDIA/TransformerEngine/blob/74ee0e0483d87d0bbf187b68f7e20b0e4632a6fc/transformer_engine/jax/sharding.py#L228) will dispatch and forward the arguments based on the sharding_type. That's why we keep those arguments but not used, it just for easier forwarding.